source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -523,14 +518,12 @@ UpstreamDrainManager::~UpstreamDrainManager() {
 }
 
 void UpstreamDrainManager::add(const Config::SharedConfigSharedPtr& config,
-                               Network::ClientConnectionPtr&& upstream_connection,
+                               Tcp::ConnectionPool::ConnectionDataPtr&& upstream_conn_data,
                                const std::shared_ptr<Filter::UpstreamCallbacks>& callbacks,
                                Event::TimerPtr&& idle_timer,
-                               const Upstream::HostDescriptionConstSharedPtr& upstream_host,
-                               Stats::TimespanPtr&& connected_timespan) {
-  DrainerPtr drainer(new Drainer(*this, config, callbacks, std::move(upstream_connection),
-                                 std::move(idle_timer), upstream_host,
-                                 std::move(connected_timespan)));
+                               const Upstream::HostDescriptionConstSharedPtr& upstream_host) {
+  DrainerPtr drainer(new Drainer(*this, config, callbacks, std::move(upstream_conn_data),
```

Comment:
Not actionable, but I wonder if the drain manager could now be moved into the cluster manager to help similar situations elsewhere? cc @ggreenway

Reply:
That's an interesting idea. I think that would work fine. It could also live in the tcp pool. That may also avoid some of the complicated lifetime-issues around stats in the current implementation.